src/test/ui/async-await/issue-98634.rs

```diff
@@ -0,0 +1,30 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+pub struct StructAsync<F: Fn() -> Pin<Box<dyn Future<Output = ()>>>> {
+    pub callback: F,
+}
+
+impl<F> Future for StructAsync<F>
+where
+    F: Fn() -> Pin<Box<dyn Future<Output = ()>>>,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        Poll::Pending
+    }
+}
+
+async fn callback() {}
+
+#[tokio::main]
```

Comment:
UI tests can't depend on non-default projects, though `tokio::main` probably isn't needed here. You can just make this a regular `async fn async_main()`, then add a blank `fn main() {}` to suppress the "no main function" error.

You can download nightly from 2022-06-27 (reported nightly version in the issue) to double check that it fails there.

Reply:
Given that the ICE mentions `MirNeighborCollector`, try calling `async_main` in your main function:

```diff
use std::{
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

pub struct StructAsync<F: Fn() -> Pin<Box<dyn Future<Output = ()>>>> {
    pub callback: F,
}

impl<F> Future for StructAsync<F>
where
    F: Fn() -> Pin<Box<dyn Future<Output = ()>>>,
{
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Pending
    }
}

async fn callback() {}

async fn async_main() {
    StructAsync { callback }.await;
}

fn main() {
+   let _ = async_main();
}
```

Reply:
If that doesn't work, you need to `cargo expand` the example and minimize away the dependency on tokio.

Reply:
With the newest change, now there is no errors (cargo run). Compilation and running successful.

Just a side note, the original code does not compile on latest nightly with a few errors.

Reply:
Minimized:

```
// edition: 2021
// check-pass

use std::{
    future::Future,
    pin::Pin,
    task::{Context, Poll, Waker},
};

pub struct StructAsync<F: Fn() -> Pin<Box<dyn Future<Output = ()>>>> {
    pub callback: F,
}

impl<F> Future for StructAsync<F>
where
    F: Fn() -> Pin<Box<dyn Future<Output = ()>>>,
{
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        Poll::Pending
    }
}

async fn callback() {}

struct Runtime;

fn waker() -> &'static Waker {
    todo!()
}

impl Runtime {
    #[track_caller]
    pub fn block_on<F: Future>(&self, mut future: F) -> F::Output {
        loop {
            unsafe {
                Pin::new_unchecked(&mut future).poll(&mut Context::from_waker(waker()));
            }
        }
    }
}

fn main() {
    Runtime.block_on(async {
        StructAsync { callback }.await;
    });
}
```

Reply:
Sorry, it *should* error, it just shouldn't *ICE*.